source/common/http/conn_manager_impl.cc

```diff
@@ -860,7 +860,9 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
 
     const bool new_metadata_added = processNewlyAddedMetadata();
-
+    if ((*entry)->end_stream_ && !new_metadata_added) {
```

Comment:
Do we need the additional `!new_metadata_added` condition here? In this case we will add a data frame and that would be covered by the buffered data check in continue, right?

Basically I'm trying to reconcile why the logic is different in the decode case vs. the encode case. Is it different because metadata is not yet implemented in the encode path? Do we need a TODO there or a comment? cc @soya3129

Reply:
sort of.  It'll work without it because the metadata path adds its own fake buffer, but it's more accurate to say we didn't send the fin with data when we know we didn't
I believe the reason we need it here and not below, is the decode path has addDecodedMetadataData and the encode path doesn't have addEncodedMetadata yet.

Reply:
Yeah, so the other route is #7720 

It amounts to a roll back of the original PR and early buffer creation.  I'm less comfortable with it because I understand what's going on less well.  That said if you are more comfortable with it we can land the fix that way, and/or could runtime (default true or false) the eager buffer creation in case it causes problems of its own.

Reply:
On the metadata encoding path, I didn't create addEncodedMetadata() because we can call callbacks_->encodeMetadata() directly ([example](https://github.com/envoyproxy/envoy/blob/78fd87b788e440574005c2f6f546d2ab8bba8dc1/test/integration/filters/response_metadata_filter.cc#L22)) to insert new metadata, and the newly added metadata will be forwarded to downstream without waiting. But I think the inconsistency can cause confusing. I will implement addEncodedMetadata().